File: Compiler/Tests/TestTokenParser.py
#!/usr/bin/env python

from TokenParser import TokenParser, ValidVariableChar
from Assert import AssertEqual, AssertTrue, AssertFalse

def TestValidVariableChar():
	tp = TokenParser('x')
	AssertTrue('tian1', ValidVariableChar('a'))
	AssertTrue('tian2', ValidVariableChar('z'))
	AssertTrue('tian3', ValidVariableChar('_'))
	AssertFalse('tian4', ValidVariableChar('?'))


def TestBasicTokens():
	tp = TokenParser("var _var1_23\n\n345   // comment\n1234{ }{} +=+-==/\n \r\n\t")
	AssertEqual('ttp1', tp.GetToken(), 'var')
	AssertEqual('ttp2', tp.GetToken(), '_var1_23')
	AssertEqual('ttp3', tp.GetToken(), '345')
	AssertEqual('ttp5', tp.GetToken(), '1234')
	AssertEqual('ttp7', tp.GetToken(), '{')
	AssertEqual('ttp8', tp.GetToken(), '}')
	AssertEqual('ttp7', tp.GetToken(), '{')
	AssertEqual('ttp8', tp.GetToken(), '}')
	AssertEqual('ttp9', tp.GetToken(), '+')
	AssertEqual('ttp10', tp.GetToken(), '=')
	AssertEqual('ttp9', tp.GetToken(), '+')
	AssertEqual('ttp11', tp.GetToken(), '-')
	AssertEqual('ttp12', tp.GetToken(), '==')
	AssertEqual('ttp13', tp.GetToken(), '/')
	AssertEqual(tp.GetToken(), None)

	AssertEqual('ttp14', tp.GetLineNumber(), 7)

def TestMoreTokens():
	tp = TokenParser("&&&")
	AssertEqual(tp.GetToken(), '&&')
	AssertEqual(tp.GetToken(), '&')

def TestOr():
	tp = TokenParser("| || |||")
	AssertEqual(tp.GetToken(), '|')
	AssertEqual(tp.GetToken(), '||')
	AssertEqual(tp.GetToken(), '||')
	AssertEqual(tp.GetToken(), '|')

def TestReplaceToken():
	tp = TokenParser('one two three')
	AssertEqual('trt1', tp.GetToken(), 'one')
	AssertEqual('trt2', tp.GetToken(), 'two')
	tp.ReplaceToken('four')
	AssertEqual('trt3', tp.GetToken(), 'four')
	AssertEqual('trt4', tp.GetToken(), 'three')

def TestScriptDefinitions():
	tp = TokenParser('scriptname(){}')
	AssertEqual(tp.GetToken(), 'scriptname')
	AssertEqual(tp.GetToken(), '(')
	AssertEqual(tp.GetToken(), ')')
	AssertEqual(tp.GetToken(), '{')
	AssertEqual(tp.GetToken(), '}')

def TestStrings():
	tp = TokenParser('bla"string"bla')
	AssertEqual(tp.GetToken(), 'bla')
	AssertEqual(tp.GetToken(), '"string"')
	AssertEqual(tp.GetToken(), 'bla')

def TestTokenParser():
	TestValidVariableChar()
	TestBasicTokens()
	TestReplaceToken()
	TestMoreTokens()
	TestOr()
	TestScriptDefinitions()
	TestStrings()

File: VC13/ScriptCompiler/ScriptCompiler/Tests/TestTokenParser.cpp


#include "..\TokenParser.h"
#include "TDDAssert.h"

class MockFileReader : public ILineFetcher
{
public:
	MockFileReader(){}
	MockFileReader(const char*script){}

private:
	virtual const char*	FetchLine()
	{
		return 0;
	}

};


void TestValidVariableChar()
{
	MockFileReader mr;
	TokenParser tp(mr);
		AssertTrue(ValidVariableChar('a'));
	AssertTrue(ValidVariableChar('z'));
	AssertTrue(ValidVariableChar('_'));
	AssertFalse(ValidVariableChar('?'));
}

void TestBasicTokens()
{
	MockFileReader mr("var _var1_23\n\n345   // comment\n1234{ }{} +=+-==/\n \r\n\t");
	TokenParser tp = TokenParser(mr);
	AssertEqual(tp.GetToken(), "var");
	AssertEqual(tp.GetToken(), "_var1_23");
	AssertEqual(tp.GetToken(), "345");
	AssertEqual(tp.GetToken(), "1234");
	AssertEqual(tp.GetToken(), "{");
	AssertEqual(tp.GetToken(), "}");
	AssertEqual(tp.GetToken(), "{");
	AssertEqual(tp.GetToken(), "}");
	AssertEqual(tp.GetToken(), "+");
	AssertEqual(tp.GetToken(), "=");
	AssertEqual(tp.GetToken(), "+");
	AssertEqual(tp.GetToken(), "-");
	AssertEqual(tp.GetToken(), "==");
	AssertEqual(tp.GetToken(), "/");
	AssertEqual(tp.GetToken(), "");

	AssertEqual(tp.GetLineNumber(), 7);
}

void TestMoreTokens()
{
	MockFileReader mr("&&&");
	TokenParser tp(mr);
	AssertEqual(tp.GetToken(), "&&");
	AssertEqual(tp.GetToken(), "&");
}

void TestOr()
{
	MockFileReader mr("| || |||");
	TokenParser tp(mr);
	AssertEqual(tp.GetToken(), "|");
	AssertEqual(tp.GetToken(), "||");
	AssertEqual(tp.GetToken(), "||");
	AssertEqual(tp.GetToken(), "|");
}

void TestReplaceToken()
{
	MockFileReader mr("one two three");
	TokenParser tp(mr);
	AssertEqual(tp.GetToken(), "one");
	AssertEqual(tp.GetToken(), "two");
	tp.ReplaceToken("four");
	AssertEqual(tp.GetToken(), "four");
	AssertEqual(tp.GetToken(), "three");
}

void TestScriptDefinitions()
{
	MockFileReader mr("scriptname(){}");
	TokenParser tp(mr);
	AssertEqual(tp.GetToken(), "scriptname");
	AssertEqual(tp.GetToken(), "(");
	AssertEqual(tp.GetToken(), ")");
	AssertEqual(tp.GetToken(), "{");
	AssertEqual(tp.GetToken(), "}");
}

void TestStrings()
{
	MockFileReader mr("bla\"string\"bla");
	TokenParser tp(mr);
	AssertEqual(tp.GetToken(), "bla");
	AssertEqual(tp.GetToken(), "\"string\"");
	AssertEqual(tp.GetToken(), "bla");
}

void TestTokenParser()
{
	TestValidVariableChar();
	TestBasicTokens();
	TestReplaceToken();
	TestMoreTokens();
	TestOr();
	TestScriptDefinitions();
	TestStrings();
}

